Detectors/CTP/workflowLumi/src/RawToLumiConverterSpec.cxx

```diff
@@ -0,0 +1,183 @@
+// See https://alice-o2.web.cern.ch/copyright for details of the copyright holders.
+// All rights not expressly granted are reserved.
+//
+// This software is distributed under the terms of the GNU General Public
+// License v3 (GPL Version 3), copied verbatim in the file "COPYING".
+//
+// In applying this license CERN does not waive the privileges and immunities
+// granted to it by virtue of its status as an Intergovernmental Organization
+// or submit itself to any jurisdiction.
+
+#include <string>
+#include "FairLogger.h"
+#include "CommonDataFormat/InteractionRecord.h"
+#include "Framework/InputRecordWalker.h"
+#include "Framework/DataRefUtils.h"
+#include "Framework/WorkflowSpec.h"
+#include "DetectorsRaw/RDHUtils.h"
+#include "DPLUtils/DPLRawParser.h"
+#include "Framework/ConfigParamRegistry.h"
+#include "CTPWorkflowLumi/RawToLumiConverterSpec.h"
+#include "CTPWorkflow/RawToDigitConverterSpec.h"
+#include "CommonUtils/VerbosityConfig.h"
+
+using namespace o2::ctp::lumi_workflow;
+
+void RawToLumiConverterSpec::init(framework::InitContext& ctx)
+{
+  mNTFToIntegrate = ctx.options().get<int>("NTF-to-IRaverage");
+}
+
+void RawToLumiConverterSpec::run(framework::ProcessingContext& ctx)
+{
+  const gbtword80_t bcidmask = 0xfff;
+  using InputSpec = o2::framework::InputSpec;
+  using ConcreteDataTypeMatcher = o2::framework::ConcreteDataTypeMatcher;
+  using Lifetime = o2::framework::Lifetime;
+  // mOutputHWErrors.clear();
+  o2::framework::DPLRawParser parser(ctx.inputs());
+  // setUpDummyLink
+  auto& inputs = ctx.inputs();
+  // if we see requested data type input with 0xDEADBEEF subspec and 0 payload this means that the "delayed message"
+  // mechanism created it in absence of real data from upstream. Processor should send empty output to not block the workflow
+  {
+    static size_t contDeadBeef = 0; // number of times 0xDEADBEEF was seen continuously
+    std::vector<InputSpec> dummy{InputSpec{"dummy", o2::framework::ConcreteDataMatcher{"CTP", "RAWDATA", 0xDEADBEEF}}};
+    for (const auto& ref : o2::framework::InputRecordWalker(inputs, dummy)) {
+      const auto dh = o2::framework::DataRefUtils::getHeader<o2::header::DataHeader*>(ref);
+      auto payloadSize = o2::framework::DataRefUtils::getPayloadSize(ref);
+      if (payloadSize == 0) {
+        auto maxWarn = o2::conf::VerbosityConfig::Instance().maxWarnDeadBeef;
+        if (++contDeadBeef <= maxWarn) {
+          LOGP(alarm, "Found input [{}/{}/{:#x}] TF#{} 1st_orbit:{} Payload {} : assuming no payload for all links in this TF{}",
+               dh->dataOrigin.str, dh->dataDescription.str, dh->subSpecification, dh->tfCounter, dh->firstTForbit, payloadSize,
+               contDeadBeef == maxWarn ? fmt::format(". {} such inputs in row received, stopping reporting", contDeadBeef) : "");
+        }
+        ctx.outputs().snapshot(o2::framework::Output{"CTP", "LUMI", 0, o2::framework::Lifetime::Timeframe}, mOutputLumiInfo);
```

Comment:
please create an empty local copy for sending dummy output, so that we know it has to be ignored.

Reply:
Sorry, I do not understand context. Please, explain a bit more.
Local copy of what ? To be ignored by what ?

Reply:
I meant that usually when detecting 0xdeadbeef we send dummy (empty) output to signal that there was no input, while you are sending the counts from the last non-empty TF. But, after second thinking, we should probably leave it as it is: since the accumulator carries the orbit of the last update, the consumer can detect that it was not updated.

Reply:
Thanks. So I keep it as it is.
For my understanding:
line https://github.com/lietava/AliceO2/blob/001cd99590ad937a693d61132abd1032dda6b50b/Detectors/CTP/workflowLumi/src/RawToLumiConverterSpec.cxx#L57
sends always last lumiInfo
while
https://github.com/lietava/AliceO2/blob/001cd99590ad937a693d61132abd1032dda6b50b/Detectors/CTP/workflow/src/RawToDigitConverterSpec.cxx#L58
send empty payload , and so you wanted me to create empty dummy lumiInfo ?

Reply:
Not sure what do you mean by L58. https://github.com/lietava/AliceO2/blob/001cd99590ad937a693d61132abd1032dda6b50b/Detectors/CTP/workflowLumi/src/RawToLumiConverterSpec.cxx#L57 is currently sending last lumiInfo in case CTP did not provide any input (hence the deadbeef was generated). Initially I was proposing to send in this case a dummy object, but then I've realized that since the lumiInfo is not really TF-specific and it has a timestamp, it is better to keep it as is. 
https://github.com/lietava/AliceO2/blob/001cd99590ad937a693d61132abd1032dda6b50b/Detectors/CTP/workflowLumi/src/RawToLumiConverterSpec.cxx#L163 sends the output updated with the last TF CTP info.